File: command/commandservice.ts
import * as discord from 'discord.js';

interface Command {
    (message: discord.Message): void;
}

class CommandService {
    register: Map<string, Command>;

    constructor() {
        this.register = new Map();
    }

    add(name: string, command: Command): void {
        this.register.set(name, command);
    }

    get(name: string): Command {
        return this.register.get(name);
    }
}

export { CommandService };

File: commands/commandService.go
package commands

import "strings"

type CommandService struct {
	register map[string]*Command
}

func NewCommandService() CommandService {
	s := CommandService{
		register: make(map[string]*Command),
	}

	return s
}

func (s *CommandService) AddCommand(command *Command) {
	s.register[command.Name] = command
}

func (s *CommandService) ExecuteCommand(ctx *Context) {
	content := strings.Trim(ctx.Message.Content, "!")
	cmd := strings.Split(content, " ")[0]

	value, ok := s.register[cmd]
	if !ok {
		return
	}

	value.Executer(ctx)
}
